Организация события

NadirCianid opened the pull request.

NadirCianid committed:
Реализовал действие "Создать событие"

NadirCianid committed:
Инкапсулировал утильные методы сервиса и контролера

NadirCianid committed:
Добавил действие "Получить событие"

NadirCianid committed:
Добавил действие "Получить все события с фильтрами"

NadirCianid committed:
Добавил действие "Удалить событие"

NadirCianid committed:
Добавил действие "Обновить событие"

NadirCianid committed:
Добавил действие "Получить все созданные пользователем события"

NadirCianid committed:
Добавил действие "Получить все события, в которых пользователь принимает участие" и причесал тесты

NadirCianid committed:
Инкапсулировал поля с дефолтным уровнем доступа в тестах мапперов

Sabina-Vishnya commented on src/test/java/school/faang/user_service/mapper/EventMapperTest.java:
```diff
@@ -0,0 +1,115 @@
+
+    @BeforeEach
+    void setUp() {
+        nadir = new User();
```
насколько помню по предыдущему пр-у у тебя похожие тестовые данные были, не только для этого класса а вообще
предлагаю сделать вот что:

1). выделить класс тестовых констант (для заполнения данными поля), чтобы не хардкодить постоянно "nadir" или еще что-то, каждый раз тебе в других тестах, где заполняешь для подготовки какой-то объект, а использовать константу просто. USER_NAME  условно будет содержать тестовое имя, ну и тоже самое к другим полям.
2). выделить класс тестовых методов, которые будут использоваться аналогичным образом. Примерно, какой-то метод createUser(USER_NAME, USER_EMAIL); будет вызываться в тестовых классах, где тебе требуется его создать. У тебя точно это не один тест, поэтому имеет смысл всё такое дублирующее вынести

Еще вариант сделать класс общий, который будет в себе содержать такие методы, константы и возможно моки какие-то еще, и наследовать его, но это может запутать и подойдет только конкретным тестам. Тут всё же без наследований обойтись пока можно, но сам посмотри.

Можно объединить это в один утилитный тестовых класс, а можно в два, или сделать класс родитель, смотри сам
Это необязательно делать в рамках текущей таски, можно завести отдельную и сделать ее после влития пр-а этого или другого, чтобы была возможность переиспользовать это в коде, который уже влили. 
Задача не сложная совсем и много времени не должна занять, по желанию можешь сделать, а можешь не делать)

У тебя и так был создан подобный класс для тестовых данных, надо лишь расширить для переиспользований в других местах.

NadirCianid replied:
Хорошо, сделаю после вливания в master

NadirCianid committed:
Инкапсулировал фильтры

NadirCianid committed:
Merge branch 'phoenix-master-bc4' into feature-BJS2-6285

NadirCianid committed:
Добавил аннотацию @Component к каждому фильтру и исправил баг с фильтрацией

NadirCianid committed:
Merge remote-tracking branch 'origin/feature-BJS2-6285' into feature-BJS2-6285

Sabina-Vishnya commented on src/main/java/school/faang/user_service/service/event/EventService.java:
```diff
@@ -0,0 +1,87 @@
+    }
+
+    public void deleteEvent(long eventId) {
+        eventRepository.delete(getEventById(eventId)
```
есть метод deleteById, не знал про него или осознанно решил не использовать?

NadirCianid replied:
Не заметил этот вариант.

Sabina-Vishnya commented on src/main/java/school/faang/user_service/service/event/EventService.java:
```diff
@@ -0,0 +1,87 @@
+        return eventRepository.findById(eventId);
+    }
+
+    public EventDto updateEvent(EventDto event) {
```
делегируешь обновлению, метод создания, да там есть валидация, но точно все случаи учтены?

Sabina-Vishnya commented on src/test/java/school/faang/user_service/service/event/EventServiceTest.java:
```diff
@@ -0,0 +1,258 @@
+
+        @DisplayName("should update event when dto is valid")
+        @Test
+        void shouldUpdateEventWhenDtoIsValid() {
```
Тестов точно один надо? если валидный, что значит валидный, кто это поймет потом?
Лучше чуть подробнее описать, тебе же будет проще спустя время к этому вернуться и другим.

На мой взгляд, было бы яснее, если бы писалось что-то типа "Когда у юзера есть навыки, связанные с событием, тогда обновить события" для позитивного, а для негативного "Когда у юзера нет навыков связанных с событием, тогда ошибка". 
А язык уже выбран английский, на нем и пишем тогда, если в команде всем будет понятно и если сам можешь, то без проблем. Это лишь как пример, с формулировкой получше можно придумать и условие точнее описать, тут лишь идея основная в том, чтобы в описании теста не просто название метода должно быть, а суть вообще того, что проверяется. 

Метод иногда бывает короче и может полностью не соответствовать описанию даже, для этого тебе как раз и стоит использовать @DisplayName, чтобы писать на человеческом языке то, что проверяет тест. В какие-то тесты иногда для описания подкидывают еще и входные данные, чтобы вывод читался. Тут кажется можем сильно не заморачиваться, а просто писать вдумчиво.
Тесты отлично помогают тебе иногда выявить баги заранее, потому что прям думаешь, что еще покрыть, что не покрыл, что вообще делает метод, а когда ты это фиксируешь это заставляет более вдумчиво код писать. Так что стоит к тестам посерьезнее относиться и оформлять их так, чтобы у коллеги, открыв тесты, вопросы отпали основные хотя бы

Sabina-Vishnya reviewed (changes requested):
предлагаю еще пока что, валидации выносить 
например, у контроллера свой класс валидации будет для событий (но обрати внимание на модификатор доступа у класса валидатора, если будешь так делать), у сервиса своя валидация может быть вынесена аналогичным образом

Еще есть вариант сделать один универсальный валидатор для null и пустоты, а еще есть аннотации и вот пока можешь просто немного посмотреть, что это. А пока предлагаю просто в отдельные классы валидаций вынести проверки. У нас так почище код будет и более декомпозированная логика

Sabina-Vishnya commented:
и + забегая вперед, погугли про ACID и @Transactional для методов работающих с бд

NadirCianid commented:
> предлагаю еще пока что, валидации выносить например, у контроллера свой класс валидации будет для событий (но обрати внимание на модификатор доступа у класса валидатора, если будешь так делать), у сервиса своя валидация может быть вынесена аналогичным образом

У меня в контроллере и в сервисе по одному методу для валидации. В сервисе он вообще в 4 строки. Как-то странно его выносить в отдельный класс. Но ладно

NadirCianid committed:
Comments fix:
- Заменил long на Long в контроллере и добавил аннотации @NonNull;
- Вынес валидацию контроллера и сервиса в отдельные методы;
- Подправил тесты под новую реализацию.

Sabina-Vishnya reviewed (commented):
- у тебя валидатор содержит часть репозиториев и маппер, по сути не сильно отличается от сервиса, в итоге, логика получается слишком размазанной, так не пойдет. Будь у тебя валидаторы разных сервисов, было бы еще понятно, а тут все в одном намешано. Оставь в валидаторе проверки полей и объектов, ты ведь можешь в сервисе обращаться к репозиториям, а в валидаторе проверять поле или объект, затем продолжать нужные тебе операции в сервисе, а если необходимо вернуть проверяемое значение, то можно ведь сделать вспомогательный метод в сервисе, затем вызывает валидатор и возвращать значение, ты и так упадешь в случае ошибки в валидаторе. В общем, идею понял, но переборщил с тем что половину логики вынес в репозиторий, попробуй не сильно логику размазывать. И вообще какие-то методы обращающиеся в другие репозитории, могут быть сервисами, аналогично как у тебя сейчас. Тогда ты вызываешь сервисы нужные, а в них уже проверки есть.
- в методе getEvent у сервиса, ты вызываешь get у Optional без предварительной проверки, а если объекта не будет? нужно проверять Optional

NadirCianid commented:
> * у тебя валидатор содержит часть репозиториев и маппер, по сути не сильно отличается от сервиса, в итоге, логика получается слишком размазанной, так не пойдет. 

я не понимаю, какая логика из валидатора не является в общем валидацией. И я не вижу здесь размазывания логики. В методах EventServiceValidation происходит только валидация без обновления базы или выполнения бизнес логики.

По сути проблема только в том, что в валидаторе я обращаюсь к репозиториям  и маперам?

Но, если запретить валидаторам обращаться к репозиториям, то останется только одна проверка - проверка на requiredSkills у владельца события. И тогда вообще теряется смысл создания целого класса с одним методом.

NadirCianid committed:
Добавил дополнительную проверку у Optional в методе EventService.getEvent()

Sabina-Vishnya commented:
> > * у тебя валидатор содержит часть репозиториев и маппер, по сути не сильно отличается от сервиса, в итоге, логика получается слишком размазанной, так не пойдет.
> 
> я не понимаю, какая логика из валидатора не является в общем валидацией. И я не вижу здесь размазывания логики. В методах EventServiceValidation происходит только валидация без обновления базы или выполнения бизнес логики.
> 
> По сути проблема только в том, что в валидаторе я обращаюсь к репозиториям и маперам?
> 
> Но, если запретить валидаторам обращаться к репозиториям, то останется только одна проверка - проверка на requiredSkills у владельца события. И тогда вообще теряется смысл создания целого класса с одним методом.

Смотри, метод getEvent вызывает сначала checkEventPresence из валидатора, в валидаторе обращение к репозиторию события и в случае отсутствия события, кидает исключение. Затем этот же getEvent вызывает приватный getEventById в сервисе, который по айди отдает событие или кидает тоже исключение. 
Тебе не кажется, что проверка одна, исключение одно, обращения к бд в итоге еще одно, без которого обойтись можно? 
Разница есть в методах, но глобально здесь можно было бы ограничиться одним методом.
Для контроллера ты вынес в валидатор и понятно, что происходит, а для сервиса ты просто вынес в новый класс валидатор, часть сервисной логики, хотя на самом деле можно было бы задуматься о том, какие методы стоит оставить, а от каких избавиться и возможно обойтись приватным методом просто или даже без него.  
В классе у тебя обращения к другим репозиториям и мапперу, почему бы не перенести часть этой логики в соответствующие сервисы? Если их нет - создать, выносить немного совсем. Ты получаешь данные и где-то мапишь - всё это необязательно должно быть именно в текущем сервисе, таким образом от части проверок ты избавишься, и это самое очевидное и не замороченное решение здесь.
Раскидывать на другие сервисы не обязательно прям в данной задаче, но тудушки хотя бы навесить и завести задачки под это можно было бы, чтобы было понимание, что ты об этом задумался. Хоть я бы делала сразу, несмотря на конфликты, которые могли бы вылезти у тебя или другого, ну значит кто-то будет их решать и избегать этого не стоит. А прежде, чем выкидывать в валидатор часть логики, подумай точно ли это подходит. И не торопись ты, лучше свежим взглядом пересмотреть потом решение и увидеть собственные недочеты, чем скорее отправлять на ревью. У тебя ведь явно код может быть понятнее и последовательнее.

Sabina-Vishnya commented:
И метод в маппере toDtos может быть без аннотации `@Mapping` (без таргета и сорса). Точно не скажу подсветит ли идея красным `"owner", target = "ownerId"`для списков, щас с телефона смотрю пр, но вообще должна, можно проверить подсветит или нет. Ну в любом случае, скомпилировать сможет. А строчку удалить можно, потому что мапстракт для списка все равно будет использовать метод toDto в цикле, а там уже маппинг есть.

NadirCianid committed:
Вынес получение скиллов в SkillService, а также удалил класс EventServiceValidation, оптимизировав проверки

Sabina-Vishnya commented:
у тебя не всё доработано и в звонке обсудили еще про это, поэтому возвращаю

NadirCianid commented:
> у тебя не всё доработано и в звонке обсудили еще про это, поэтому возвращаю

не понимаю, что именно не доработано + на звонке мы обсуждали другу мою задачу. В данной задаче у меня нет сторонних классов в слое сервиса, кроме самого сервиса

NadirCianid committed:
Переименовал метода EventService.checkOwnerSkills на EventService.doesOwnerHaveEventRelatedSkills

Sabina-Vishnya commented:
> > у тебя не всё доработано и в звонке обсудили еще про это, поэтому возвращаю
> 
> не понимаю, что именно не доработано + на звонке мы обсуждали другу мою задачу. В данной задаче у меня нет сторонних классов в слое сервиса, кроме самого сервиса

да, я видать тогда не тот коммит глянула, всё гуд)
